pkg/controller/directvolumemigration/rsync.go

```diff
@@ -1651,3 +1678,591 @@ func (t *Task) deleteProgressReportingCRs(client k8sclient.Client) error {
+			return true
+		}
+	}
+	return false
+}
+
+// Failed returns number of failed operations
+func (r *rsyncClientOperationStatusList) Failed() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.failed {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Succeeded returns number of failed operations
+func (r *rsyncClientOperationStatusList) Succeeded() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.succeeded {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Pending returns number of pending operations
+func (r *rsyncClientOperationStatusList) Pending() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.pending {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Running returns number of running operations
+func (r *rsyncClientOperationStatusList) Running() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.pending {
+			i += 1
+		}
+	}
+	return i
+}
+
+// ensureRsyncOperations orchestrates all attempts of Rsync, updates owner status with observed state of all ongoing Rsync operations
+// returns structured status of all operations, the return value should be used to make decisions about whether to retry in next reconcile
+func (t *Task) ensureRsyncOperations(client compat.Client, podRequirements []rsyncClientPodRequirements) (rsyncClientOperationStatusList, error) {
+	statusList := rsyncClientOperationStatusList{}
+	// rateLimiter defines maximum concurrent operations that can be reconciled in one go
+	rateLimiter := make(chan bool, DefaultRsyncOperationConcurrency)
+	// outputChan streamlines output of concurrent reconcile operations
+	outputChan := make(chan rsyncClientOperationStatus, DefaultRsyncOperationConcurrency)
+	finishChan := make(chan bool)
+	waitGroup := &sync.WaitGroup{}
+	mutex := &sync.Mutex{}
+	// when garbage collection is enabled, delete intermittent Rsync pods
+	// TODO: implement garbage collection when needed
+	// go t.garbageCollectPodsForRequirements(podRequirements)
+	for i := range podRequirements {
+		req := &podRequirements[i]
+		lastObservedOperationStatus := t.Owner.Status.GetRsyncOperationStatusForPVC(&corev1.ObjectReference{
+			Name:      req.pvInfo.name,
+			Namespace: req.namespace,
+		})
+		// if the Rsync operation is already completed, do nothing
+		if lastObservedOperationStatus.IsComplete() {
+			continue
+		}
+		t.reconcileRsyncOperationState(
+			client,
+			req,
+			lastObservedOperationStatus,
+			rateLimiter,
+			outputChan,
+			waitGroup)
+	}
+	// consume output from concurrent operations
+	go func() {
+		for incomingOutput := range outputChan {
+			mutex.Lock()
+			statusList.Add(incomingOutput)
+			mutex.Unlock()
+		}
+		finishChan <- true
+	}()
+	waitGroup.Wait()
+	close(rateLimiter)
+	close(outputChan)
+	<-finishChan
+	return statusList, nil
+}
+
+// garbageCollectRsyncPods garbage collection routine
+// will run in background, returns list of failures when encountered
+func (t *Task) garbageCollectPodsForRequirements(podRequirements []rsyncClientPodRequirements) []error {
```

Comment:
not implemented yet. Two concerns to consider: 
1. Where should this happen? Want to make sure the reconcile is fast enough
2. If we fail to delete a pod, do we just report errors as they happen and continue deleting next pod making it at best effort?